src/zone/napa-zone.cpp

```diff
@@ -78,25 +85,49 @@ namespace {
     struct ZoneData;
 
 
-    struct NapaZoneEventEmitter {
-        // TBD: NapaZoneEventEmitter: the following methods should be implemented.
-        // They should run very fast (no any blocking operation)
-        void OnCreated() {}
-        void OnRecycling() {}
-        void OnRecycled() {}
-        void OnTerminated() {}
-
-        ~NapaZoneEventEmitter() {
-            NAPA_DEBUG("Zone", "Destructor NapaZoneEventEmitter");
-        }
-    };
-
+  
     struct NapaZoneImpl { // address of 'this' will be stored to TLS (WorkerContextItem::ZONE)
         Zone::State _state;
         settings::ZoneSettings _settings; // address of '_settings.id' will be stored to TLS (WorkerContextItem::ZONE_ID)
         std::unique_ptr<Scheduler> _scheduler;
         std::shared_ptr<ZoneData> _zoneData;
-        NapaZoneEventEmitter _events;
+
+        // run emitter in node's main thread with parameters
+        void EmitEvent(const char* event, ...) {
```

Comment:
I strongly recommend to use strong types:
```
typedef (void *ZoneRecycledCallback)();
typedef (void *ZoneRecyclingCallback)();
typedef (void *ZoneExitCallback)(napa::zone_exit_code);
```

dynamic semantics are not helpful here in native code -- string keys for event type and variadic arguments.

Reply:
I prefer simplicity here rather than over design. All these things are internal used in this single file!

---
In reply to: [186880299](https://github.com/helloshuangzi/napajs/pull/20#discussion_r186880299) [](ancestors = 186880299)

Reply:
This is not about over-design or not.. Let me clarify 2 points:
 - use string as key of event type is not necessary here in native code. Suggestion: use different function, or template, or enum. (enum is not good enough comparing to other solutions, but better than using string)
 - use variadic arguments should be avoid here. The numbers of arguments are determined in build-time. Suggestion: use template specification.

The key point is, do not introduce unnecessary dynamic bindings in native code.